### ISSUE TITLE
Checkpointer crashes on `jsonplus` with StructuredTools

### ISSUE THREAD

**Issue by michael-rubel** [langchain-ai/langgraph#1963]
### Checked other resources

- [X] I added a very descriptive title to this issue.
- [X] I searched the [LangGraph](https://langchain-ai.github.io/langgraph/)/LangChain documentation with the integrated search.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I am sure that this is a bug in LangGraph/LangChain rather than my code.
- [X] I am sure this is better as an issue [rather than a GitHub discussion](https://github.com/langchain-ai/langgraph/discussions/new/choose), since this is a LangGraph bug and not a design question.

### Example Code

```python
tools = [
    StructuredTool.from_function(
         func=self.customer.find_customer,
         name='find_customer',
         description='Description',
    ),
    ... other tools like this one ^
]

memory = SqliteSaver(conn=sqlite3_conn)

create_react_agent(
   model=openai,
   tools=tools,
   state_modifier=prompt,
   checkpointer=memory,
)
```

### Error Message and Stack Trace (if applicable)

```shell
Traceback (most recent call last):
File "/home/michael/PycharmProjects/oro-ai/.venv/lib/python3.11/site-packages/langgraph/pregel/__init__.py", line 1560, in invoke
    for chunk in self.stream(
  File "/home/michael/PycharmProjects/oro-ai/.venv/lib/python3.11/site-packages/langgraph/pregel/__init__.py", line 1248, in stream
    with SyncPregelLoop(
  File "/home/michael/PycharmProjects/oro-ai/.venv/lib/python3.11/site-packages/langgraph/pregel/loop.py", line 770, in __exit__
    return self.stack.__exit__(exc_type, exc_value, traceback)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/michael/.pyenv/versions/3.11.9/lib/python3.11/contextlib.py", line 601, in __exit__
    raise exc_details[1]
  File "/home/michael/.pyenv/versions/3.11.9/lib/python3.11/contextlib.py", line 586, in __exit__
    if cb(*exc_details):
       ^^^^^^^^^^^^^^^^
  File "/home/michael/PycharmProjects/oro-ai/.venv/lib/python3.11/site-packages/langgraph/pregel/executor.py", line 105, in __exit__
    task.result()
  File "/home/michael/.pyenv/versions/3.11.9/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/home/michael/.pyenv/versions/3.11.9/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/home/michael/PycharmProjects/oro-ai/.venv/lib/python3.11/site-packages/langgraph/pregel/executor.py", line 70, in done
    task.result()
  File "/home/michael/.pyenv/versions/3.11.9/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/home/michael/.pyenv/versions/3.11.9/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/home/michael/.pyenv/versions/3.11.9/lib/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/michael/PycharmProjects/oro-ai/.venv/lib/python3.11/site-packages/langgraph/pregel/loop.py", line 696, in _checkpointer_put_after_previous
    prev.result()
  File "/home/michael/.pyenv/versions/3.11.9/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/home/michael/.pyenv/versions/3.11.9/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/home/michael/PycharmProjects/oro-ai/.venv/lib/python3.11/site-packages/langgraph/pregel/executor.py", line 70, in done
    task.result()
  File "/home/michael/.pyenv/versions/3.11.9/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/home/michael/.pyenv/versions/3.11.9/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/home/michael/.pyenv/versions/3.11.9/lib/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/michael/PycharmProjects/oro-ai/.venv/lib/python3.11/site-packages/langgraph/pregel/loop.py", line 696, in _checkpointer_put_after_previous
    prev.result()
  File "/home/michael/.pyenv/versions/3.11.9/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/home/michael/.pyenv/versions/3.11.9/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/home/michael/PycharmProjects/oro-ai/.venv/lib/python3.11/site-packages/langgraph/pregel/executor.py", line 70, in done
    task.result()
  File "/home/michael/.pyenv/versions/3.11.9/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/home/michael/.pyenv/versions/3.11.9/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/home/michael/.pyenv/versions/3.11.9/lib/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/michael/PycharmProjects/oro-ai/.venv/lib/python3.11/site-packages/langgraph/pregel/loop.py", line 698, in _checkpointer_put_after_previous
    cast(BaseCheckpointSaver, self.checkpointer).put(
  File "/home/michael/PycharmProjects/oro-ai/.venv/lib/python3.11/site-packages/langgraph/checkpoint/sqlite/__init__.py", line 399, in put
    type_, serialized_checkpoint = self.serde.dumps_typed(checkpoint)
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/michael/PycharmProjects/oro-ai/.venv/lib/python3.11/site-packages/langgraph/checkpoint/serde/jsonplus.py", line 193, in dumps_typed
    return "msgpack", _msgpack_enc(obj)
                      ^^^^^^^^^^^^^^^^^
  File "/home/michael/PycharmProjects/oro-ai/.venv/lib/python3.11/site-packages/langgraph/checkpoint/serde/jsonplus.py", line 500, in _msgpack_enc
    return enc.pack(data)
           ^^^^^^^^^^^^^^
  File "msgpack/_packer.pyx", line 279, in msgpack._cmsgpack.Packer.pack
  File "msgpack/_packer.pyx", line 276, in msgpack._cmsgpack.Packer.pack
  File "msgpack/_packer.pyx", line 265, in msgpack._cmsgpack.Packer._pack
  File "msgpack/_packer.pyx", line 213, in msgpack._cmsgpack.Packer._pack_inner
  File "msgpack/_packer.pyx", line 265, in msgpack._cmsgpack.Packer._pack
  File "msgpack/_packer.pyx", line 213, in msgpack._cmsgpack.Packer._pack_inner
  File "msgpack/_packer.pyx", line 265, in msgpack._cmsgpack.Packer._pack
  File "msgpack/_packer.pyx", line 213, in msgpack._cmsgpack.Packer._pack_inner
  File "msgpack/_packer.pyx", line 265, in msgpack._cmsgpack.Packer._pack
  File "msgpack/_packer.pyx", line 232, in msgpack._cmsgpack.Packer._pack_inner
  File "msgpack/_packer.pyx", line 265, in msgpack._cmsgpack.Packer._pack
  File "msgpack/_packer.pyx", line 213, in msgpack._cmsgpack.Packer._pack_inner
  File "msgpack/_packer.pyx", line 267, in msgpack._cmsgpack.Packer._pack
  File "/home/michael/PycharmProjects/oro-ai/.venv/lib/python3.11/site-packages/langgraph/checkpoint/serde/jsonplus.py", line 232, in _msgpack_default
    obj.model_dump(),
    ^^^^^^^^^^^^^^^^
TypeError: BaseModel.model_dump() missing 1 required positional argument: 'self'
```

### Description

SqliteSaver fails with create_react_agent() construct and `StructuredTools`

Actually, the same happens with the basic `MemorySaver()`

Versions:
langchain = "^0.3"
langchain-cli = "^0.0.31"
langchain-openai = "^0.2"
langgraph = "^0.2"
langgraph-checkpoint-sqlite = "^2.0.0"

### System Info

System Information
------------------
> OS:  Linux
> OS Version:  #26-Ubuntu SMP PREEMPT_DYNAMIC Tue Jun 18 12:35:22 UTC 2024
> Python Version:  3.11.9 (main, Aug 12 2024, 07:24:05) [GCC 11.4.0]

Package Information
-------------------
> langchain_core: 0.3.7
> langchain: 0.3.1
> langchain_community: 0.3.1
> langsmith: 0.1.129
> langchain_cli: 0.0.31
> langchain_openai: 0.2.1
> langchain_text_splitters: 0.3.0
> langgraph: 0.2.32
> langserve: 0.3.0

Other Dependencies
------------------
> aiohttp: 3.10.8
> async-timeout: Installed. No version info available.
> dataclasses-json: 0.6.7
> fastapi: 0.115.0
> gitpython: 3.1.43
> gritql: 0.1.5
> httpx: 0.27.2
> jsonpatch: 1.33
> langgraph-checkpoint: 2.0.0
> langserve[all]: Installed. No version info available.
> numpy: 1.26.4
> openai: 1.51.0
> orjson: 3.10.7
> packaging: 24.1
> pydantic: 2.9.2
> pydantic-settings: 2.5.2
> PyYAML: 6.0.2
> requests: 2.32.3
> SQLAlchemy: 2.0.35
> sse-starlette: 1.8.2
> tenacity: 8.4.2
> tiktoken: 0.7.0
> tomlkit: 0.12.5
> typer[all]: Installed. No version info available.
> typing-extensions: 4.12.2
> uvicorn: 0.23.2

**Comment by gbaian10** [comment]
Hi, can you provide a more complete example?
For example: Testing it like this is normal.

```py
from dotenv import load_dotenv
from langchain_core.tools import StructuredTool
from langchain_openai import ChatOpenAI
from langgraph.checkpoint.memory import MemorySaver
from langgraph.prebuilt import create_react_agent

load_dotenv()


class MyTool:
    def add(self, a: int, b: int) -> int:
        """Add two numbers."""
        return a + b

    def get_tools(self) -> list[StructuredTool]:
        return [
            StructuredTool.from_function(
                func=self.add,
                name="find_customer",
                description="Description",
            ),
        ]


openai = ChatOpenAI(model="gpt-4o-mini")
app = create_react_agent(
    model=openai, tools=MyTool().get_tools(), checkpointer=MemorySaver()
)
thread_config = {"configurable": {"thread_id": "1"}}
print(app.invoke({"messages": "What is 1 + 1?"}, thread_config))
```

I'm not sure about your `func=self.customer.find_customer`, in what context it is used.

**Comment by gbaian10** [comment]
Ah, it didn't call the tool. I'll test it again.

**Comment by michael-rubel** [comment]
I found the reason.

I was passing additional arguments from typed object to the `invoke` method dynamically, like this:
```
runnable.invoke(params.model_dump(), , config={'configurable': {
    'thread_id': params.thread_id,
}})
```
And when I started using the checkpointer it started failing.

Note: It may be worth adding to the docs that passing additional stuff might harm the executor when the checkpointer is utilized.

Thanks @gbaian10

**Comment by VKudlay** [comment]
Ran into a similar issue. The same error can happen if you try to pass around a schema class in your state (i.e. as a variable, in a dict, etc.), since the check will see it has the `model_dump` method and will attempt to call it. The simple workaround in my case was to just construct an instance of the schema and then pass that around instead; I only cared about the class methods, so it was needless but easy enough. [Relevant line to fix:](https://github.com/langchain-ai/langgraph/blob/1ef71211003ed95190079056baf0cf3026c6eec0/libs/checkpoint/langgraph/checkpoint/serde/jsonplus.py#L227) 
```python
def _msgpack_default(obj: Any) -> Union[str, msgpack.ExtType]:
    if hasattr(obj, "model_dump") and callable(obj.model_dump):  # pydantic v2  ## <- insufficient check
        return msgpack.ExtType(
            EXT_PYDANTIC_V2,
            _msgpack_enc(
                (
                    obj.__class__.__module__,
                    obj.__class__.__name__,
                    obj.model_dump(), ## <- error line
                    "model_validate_json",
                ),
            ),
        )
```

It doesn't sound like something most people run into and is easy to work around. Changing `callable` check to `isinstance` sounds obvious but might break some fringe use-cases, so I'll just leave it alone and park this post as a lighthouse.
